text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+(see [this protobuf specification](#appendix-a---protocol-buffer-definitions) for more details) exposing a single API endpoint named `ArrowStream`.
+This endpoint is based on a bidirectional streaming protocol. The ingress side is a `BatchArrowRecords` stream encoding a
+batch of Apache Arrow buffers (more specifically [Arrow IPC format](#arrow-ipc-format)). The egress
+side is a `BatchStatus` stream reporting asynchronously the status of each `BatchArrowRecords` previously sent.
+
+After establishing the underlying transport the client starts sending telemetry data using the `ArrowStream` request.
+The
+client continuously sends `BatchArrowRecords`'s messages over the opened stream to the server and expects to receive
+continuously
+`BatchStatus`'s messages from the server as illustrated by the following sequence diagram:
+
+![Sequence diagram](img/0156_OTEL%20-%20ProtocolSeqDiagram.png)
+
+> Multiple streams can be simultaneously opened between a client and a server to increase the maximum achievable
+throughput.
+
+If the client is shutting down (e.g. when the containing process wants to exit) the client will optionally wait until
+all pending acknowledgements are received or until an implementation specific timeout expires. This ensures reliable
+delivery of telemetry data. The client implementation should expose an option to turn on and off the waiting during shutdown.
+This behavior is described in more details in the section [Best effort delivery guarantee](#best-effort-delivery-guarantee).
+
+The protobuf definition of this service is:
+
+```protobuf
+service ArrowStreamService {
+  // The ArrowStream endpoint is a bi-directional stream used to send batch of `BatchArrowRecords` from the exporter
+  // to the collector. The collector returns `BatchStatus` messages to acknowledge the `BatchArrowRecords` messages received.
+  rpc ArrowStream(stream BatchArrowRecords) returns (stream BatchStatus) {}
+}
+```
+
+> **Unary RPC vs Stream RPC**: We use a stream-oriented protocol **to get rid of the overhead of specifying the schema
+> and dictionaries for each batch.** A state will be maintained receiver side to keep track of the schemas and
+> dictionaries. The [Arrow IPC format](#arrow-ipc-format) has been designed to follow this pattern and also allows the
+> dictionaries to be sent incrementally. Similarly, ZSTD dictionaries can also be transferred to the RPC stream to
+> optimize the transfer of small batches (for more details see the description of the compression field in the
+> following paragraphs). To mitigate the usual pitfalls of a stream-oriented protocol please see this [paragraph](#traffic-balancing-optimization) in the
+> implementation recommendations section.
+
+A `BatchArrowRecords` message is composed of 3 attributes. The protobuf definition is:
+
+```protobuf
+message BatchArrowRecords {
+  // [mandatory] Batch ID. Must be unique in the context of the stream.
+  string batch_id = 1;
```

Comment:
Why string and not e.g. 64 bit integer sequence number?

Reply:
Good question -- I think this can change to integer.